internal/rust/semantic/src/types.rs

```diff
@@ -0,0 +1,666 @@
+use std::{cmp, collections::HashMap, fmt};
+
+// PolyType represents a generic parametrized type.
+//
+// TODO:
+//     Do not derive PartialEq implementation.
+//     Instead provide a custom implementation
+//     that instantiates both polytypes with the
+//     same type variables.
+//
+//     Note this requires a substitution, so remove
+//     this derivation once substitutions are defined.
+//
+#[derive(Debug, Clone, PartialEq)]
+pub struct PolyType {
+    pub free: Vec<Tvar>,
+    pub bnds: Option<HashMap<Tvar, Kind>>,
+    pub expr: MonoType,
+}
+
+impl fmt::Display for PolyType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &self.bnds {
+            Some(bnds) => {
+                let mut bounds = Vec::new();
+                for tv in &self.free {
+                    if let Some(kind) = bnds.get(tv) {
+                        bounds.push(BoundTvar {
+                            tv: *tv,
+                            kind: *kind,
+                        })
+                    }
+                }
+                write!(
+                    f,
+                    "forall [{}] where {} {}",
+                    List { values: &self.free },
+                    List { values: &bounds },
+                    self.expr,
+                )
+            }
+            None => write!(f, "forall [{}] {}", List { values: &self.free }, self.expr),
+        }
+    }
+}
+
+// Kind represents a class or family of types
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Kind {
+    Add,
+    Div,
+    Cmp,
+    Nul,
+}
+
+impl fmt::Display for Kind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Kind::Add => f.write_str("Addable"),
+            Kind::Div => f.write_str("Divisible"),
+            Kind::Cmp => f.write_str("Comparable"),
+            Kind::Nul => f.write_str("Nullable"),
+        }
+    }
+}
+
+// MonoType represents a specific named type
+#[derive(Debug, Clone, PartialEq)]
+pub enum MonoType {
+    Bool,
+    Int,
+    Uint,
+    Flt,
+    Str,
+    Dur,
+    Time,
+    Rgx,
```

Comment:
Similar question here, I would rather spell these all the way out, i.e. `Float`, `String` and `Regexp`.  `Uint` is fine as that is very common.